File: tests/cli/integration/gate-matrix.test.ts
/**
 * E7 fitness function: gate state machine matrix (NON-INTERACTIVE rows).
 *
 * The drift gate composes three R21–R24 choices (discard / persist / abort)
 * with two execution modes (interactive / non-interactive). The full
 * documented matrix is 3×2, but this file only exercises the
 * **non-interactive** row at the binary surface — that's what the spawn
 * runner can drive without a pseudo-TTY:
 *
 *   |                 | discard | persist | abort |
 *   | non-interactive |    ✓    |    ✓    |   ✓   |
 *   | interactive     |    —    |    —    |   —   |   ← unit-level only
 *
 * Interactive cells are covered at the unit level in tests/drift/gate.test.ts
 * (decideGate has interactive vs non-interactive branches with stub prompts);
 * a binary-surface pseudo-TTY harness is deferred (would need node-pty).
 *
 * What this matrix DOES verify across the spawn boundary:
 *   - non-interactive without --on-drift: exit 1 + flag mentioned
 *   - non-interactive --on-drift=discard: exit 0 + new content + backup snapshot
 *   - non-interactive --on-drift=persist: exit 0 + active profile updated + new content
 *   - non-interactive --on-drift=abort: exit 1 + state unchanged
 *   - sync (active profile re-materialise) under each --on-drift= choice
 */

import { promises as fs } from "node:fs";
import * as path from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { merge } from "../../../src/merge/merge.js";
import { resolve } from "../../../src/resolver/resolve.js";
import { materialize } from "../../../src/state/materialize.js";
import { buildStatePaths } from "../../../src/state/paths.js";
import { makeFixture, type Fixture } from "../../helpers/fixture.js";

import { ensureBuilt, runCli } from "./spawn.js";

let fx: Fixture | undefined;
afterEach(async () => {
  if (fx) await fx.cleanup();
  fx = undefined;
});

/** Two-profile fixture with `a` active and `.claude/CLAUDE.md` drifted to "EDIT\n". */
async function setupDrifted() {
  const f = await makeFixture({
    profiles: {
      a: { manifest: { name: "a" }, files: { "CLAUDE.md": "A\n" } },
      b: { manifest: { name: "b" }, files: { "CLAUDE.md": "B\n" } },
    },
  });
  const planA = await resolve("a", { projectRoot: f.projectRoot });
  const m = await merge(planA);
  await materialize(buildStatePaths(f.projectRoot), planA, m);
  await fs.writeFile(path.join(f.projectRoot, ".claude", "CLAUDE.md"), "EDIT\n");
  return f;
}

describe("E7 gate state machine matrix — non-interactive cells", () => {
  it("non-interactive without --on-drift → exit 1 + stderr names the flag", async () => {
    await ensureBuilt();
    fx = await setupDrifted();
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "use", "b"],
    });
    expect(r.exitCode).toBe(1);
    expect(r.stderr).toContain("--on-drift=");
    // Live tree is unchanged (the abort precedes any write).
    expect(
      await fs.readFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "utf8"),
    ).toBe("EDIT\n");
  });

  it("non-interactive --on-drift=discard → exit 0 + new content + backup snapshot equals pre-swap edited tree", async () => {
    await ensureBuilt();
    fx = await setupDrifted();
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--on-drift=discard", "use", "b"],
    });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toContain("Switched to b");
    // Backup snapshot must be on disk under .claude-profiles/.backup/, and
    // the snapshot bytes must equal the drifted live tree at the moment of
    // discard (R23a — pre-swap snapshot, not the post-swap content).
    const backupDir = path.join(fx.projectRoot, ".claude-profiles", ".backup");
    // Filter to directories and sort by name (snapshot dirs are timestamped
    // and lexicographically sortable). A stray file in .backup/ must not
    // confuse the lookup; the production prune path applies the same
    // filter (see src/state/backup.ts:pruneOldSnapshots).
    const dirents = await fs.readdir(backupDir, { withFileTypes: true });
    const snapshotDirs = dirents
      .filter((d) => d.isDirectory())
      .map((d) => d.name)
      .sort();
    expect(snapshotDirs.length).toBeGreaterThan(0);
    const snapshot = await fs.readFile(
      path.join(backupDir, snapshotDirs[0]!, "CLAUDE.md"),
      "utf8",
    );
    expect(snapshot).toBe("EDIT\n");
    // Live tree now matches b.
    expect(
      await fs.readFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "utf8"),
    ).toBe("B\n");
  });

  it("non-interactive --on-drift=persist → exit 0 + drift saved to active profile + new content live", async () => {
    await ensureBuilt();
    fx = await setupDrifted();
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--on-drift=persist", "use", "b"],
    });
    expect(r.exitCode).toBe(0);
    // The previously-active profile a now has the edited content baked in.
    expect(
      await fs.readFile(
        path.join(fx.projectRoot, ".claude-profiles", "a", ".claude", "CLAUDE.md"),
        "utf8",
      ),
    ).toBe("EDIT\n");
    // Live tree is now b's content.
    expect(
      await fs.readFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "utf8"),
    ).toBe("B\n");
  });

  it("non-interactive --on-drift=abort → exit 1 + state and live tree unchanged", async () => {
    await ensureBuilt();
    fx = await setupDrifted();
    const stateBefore = await fs.readFile(
      path.join(fx.projectRoot, ".claude-profiles", ".state.json"),
      "utf8",
    );
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--on-drift=abort", "use", "b"],
    });
    expect(r.exitCode).toBe(1);
    // .claude/ untouched
    expect(
      await fs.readFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "utf8"),
    ).toBe("EDIT\n");
    // .state.json untouched
    const stateAfter = await fs.readFile(
      path.join(fx.projectRoot, ".claude-profiles", ".state.json"),
      "utf8",
    );
    expect(stateAfter).toBe(stateBefore);
  });
});

describe("E7 gate state machine matrix — sync (re-materialise active) cells", () => {
  it("sync without --on-drift on a drifted active → exit 1 (S12 gate)", async () => {
    await ensureBuilt();
    fx = await setupDrifted();
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "sync"] });
    expect(r.exitCode).toBe(1);
    expect(r.stderr).toContain("--on-drift=");
  });

  it("sync --on-drift=discard on drifted active → exit 0 + live restored to source", async () => {
    await ensureBuilt();
    fx = await setupDrifted();
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--on-drift=discard", "sync"],
    });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toContain("Synced a");
    expect(
      await fs.readFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "utf8"),
    ).toBe("A\n");
  });

  it("sync --on-drift=abort on drifted active → exit 1 + live tree unchanged", async () => {
    await ensureBuilt();
    fx = await setupDrifted();
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "--on-drift=abort", "sync"],
    });
    expect(r.exitCode).toBe(1);
    expect(
      await fs.readFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "utf8"),
    ).toBe("EDIT\n");
  });

  it("sync on clean active → no-op success (no drift to gate on)", async () => {
    await ensureBuilt();
    const f = await makeFixture({
      profiles: {
        a: { manifest: { name: "a" }, files: { "CLAUDE.md": "A\n" } },
      },
    });
    fx = f;
    const planA = await resolve("a", { projectRoot: f.projectRoot });
    const m = await merge(planA);
    await materialize(buildStatePaths(f.projectRoot), planA, m);
    const r = await runCli({ args: ["--cwd", f.projectRoot, "sync"] });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toContain("Synced a");
  });
});

File: go/tests/integration/gate_matrix_test.go
package integration_test

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"

	"github.com/htxryan/c3p/tests/integration/helpers"
)

// TestGateMatrix — IV/T3 translation of TS gate-matrix.test.ts.
//
// E7 fitness function: the drift gate state machine for non-interactive
// rows. The full documented matrix is 3x2 (discard/persist/abort x
// interactive/non-interactive); this file exercises the non-interactive
// row + sync re-materialise paths at the binary surface. Interactive
// branches are unit-level only (would need a pty harness).
//
// TS used in-process imports (resolve + merge + materialize) to set up
// "active = a, .claude/ drifted to EDIT". The Go translation drives the
// same setup through the CLI: `use a` materialises, then we write the
// drift directly. No internal Go imports.

// setupDrifted creates a two-profile fixture with `a` active and
// .claude/CLAUDE.md drifted to "EDIT\n". Mirrors the TS helper.
func setupDrifted(t *testing.T) *helpers.Fixture {
	t.Helper()
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"a": {Manifest: map[string]any{"name": "a"}, Files: map[string]string{"CLAUDE.md": "A\n"}},
			"b": {Manifest: map[string]any{"name": "b"}, Files: map[string]string{"CLAUDE.md": "B\n"}},
		},
	})
	// Materialise `a` via the CLI.
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "a"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("setup use a: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	// Drift the live tree.
	if err := os.WriteFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"), []byte("EDIT\n"), 0o644); err != nil {
		t.Fatalf("write drift: %v", err)
	}
	return fx
}

// TestGateMatrix_NonInteractive_NoFlag — without --on-drift, exit 1 +
// stderr names the flag; live tree unchanged.
func TestGateMatrix_NonInteractive_NoFlag(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDrifted(t)
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "b"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("use b no flag: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "--on-drift=") {
		t.Errorf("stderr missing --on-drift= guidance: %q", r.Stderr)
	}
	got, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read live: %v", err)
	}
	if string(got) != "EDIT\n" {
		t.Errorf("live tree mutated: want EDIT, got %q", got)
	}
}

// TestGateMatrix_NonInteractive_Discard — --on-drift=discard: exit 0 +
// new content live + backup snapshot equals pre-swap edited tree
// (R23a — pre-swap snapshot, not post-swap content).
func TestGateMatrix_NonInteractive_Discard(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDrifted(t)
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--on-drift=discard", "use", "b"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("use b discard: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	// Go bin lowercases verb: "switched to b" (TS: "Switched to b").
	if !strings.Contains(strings.ToLower(r.Stdout), "switched to b") {
		t.Errorf("stdout missing 'switched to b': %q", r.Stdout)
	}
	// Backup snapshot directory must contain at least one timestamped dir,
	// and the snapshot bytes must equal the pre-swap drifted tree.
	backupDir := filepath.Join(fx.ProjectRoot, ".claude-profiles", ".meta", "backup")
	dirents, err := os.ReadDir(backupDir)
	if err != nil {
		t.Fatalf("read backup dir: %v", err)
	}
	var snapshotDirs []string
	for _, d := range dirents {
		if d.IsDir() {
			snapshotDirs = append(snapshotDirs, d.Name())
		}
	}
	if len(snapshotDirs) == 0 {
		t.Fatalf("no snapshot dir under %s", backupDir)
	}
	sort.Strings(snapshotDirs)
	snapshot, err := os.ReadFile(filepath.Join(backupDir, snapshotDirs[0], "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read snapshot: %v", err)
	}
	if string(snapshot) != "EDIT\n" {
		t.Errorf("snapshot bytes: want EDIT, got %q", snapshot)
	}
	// Live tree now matches b.
	live, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read live: %v", err)
	}
	if string(live) != "B\n" {
		t.Errorf("live tree: want B, got %q", live)
	}
}

// TestGateMatrix_NonInteractive_Persist — --on-drift=persist: exit 0 +
// drift saved to active profile + new content live.
func TestGateMatrix_NonInteractive_Persist(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDrifted(t)
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--on-drift=persist", "use", "b"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("use b persist: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	// Previously-active profile a now has the edited content baked in.
	persisted, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude-profiles", "a", ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read persisted a: %v", err)
	}
	if string(persisted) != "EDIT\n" {
		t.Errorf("persisted a: want EDIT, got %q", persisted)
	}
	live, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read live: %v", err)
	}
	if string(live) != "B\n" {
		t.Errorf("live: want B, got %q", live)
	}
}

// TestGateMatrix_NonInteractive_Abort — --on-drift=abort: exit 1 +
// state.json and live tree unchanged.
func TestGateMatrix_NonInteractive_Abort(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDrifted(t)
	statePath := filepath.Join(fx.ProjectRoot, ".claude-profiles", ".meta", "state.json")
	stateBefore, err := os.ReadFile(statePath)
	if err != nil {
		t.Fatalf("read state.json before: %v", err)
	}
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--on-drift=abort", "use", "b"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("use b abort: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	live, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read live: %v", err)
	}
	if string(live) != "EDIT\n" {
		t.Errorf("live tree mutated: want EDIT, got %q", live)
	}
	stateAfter, err := os.ReadFile(statePath)
	if err != nil {
		t.Fatalf("read state.json after: %v", err)
	}
	if string(stateAfter) != string(stateBefore) {
		t.Errorf("state.json mutated under abort:\nbefore: %q\nafter:  %q", stateBefore, stateAfter)
	}
}

// TestGateMatrix_Sync_NoFlag — sync on a drifted active without
// --on-drift → exit 1 (S12 gate).
func TestGateMatrix_Sync_NoFlag(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDrifted(t)
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "sync"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("sync no flag: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "--on-drift=") {
		t.Errorf("stderr missing --on-drift= guidance: %q", r.Stderr)
	}
}

// TestGateMatrix_Sync_Discard — sync --on-drift=discard on drifted
// active → exit 0 + live restored to source.
func TestGateMatrix_Sync_Discard(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDrifted(t)
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--on-drift=discard", "sync"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("sync discard: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	// Go bin: "synced a" (lowercase). TS bin used "Synced a".
	if !strings.Contains(strings.ToLower(r.Stdout), "synced a") {
		t.Errorf("stdout missing 'synced a': %q", r.Stdout)
	}
	live, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read live: %v", err)
	}
	if string(live) != "A\n" {
		t.Errorf("live: want A, got %q", live)
	}
}

// TestGateMatrix_Sync_Abort — sync --on-drift=abort → exit 1 + live
// tree unchanged.
func TestGateMatrix_Sync_Abort(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupDrifted(t)
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "--on-drift=abort", "sync"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("sync abort: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	live, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read live: %v", err)
	}
	if string(live) != "EDIT\n" {
		t.Errorf("live tree mutated: want EDIT, got %q", live)
	}
}

// TestGateMatrix_Sync_CleanActive — sync on a clean active is a no-op
// success (no drift → no gate).
func TestGateMatrix_Sync_CleanActive(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"a": {Manifest: map[string]any{"name": "a"}, Files: map[string]string{"CLAUDE.md": "A\n"}},
		},
	})
	// Materialise a via the CLI (no drift → clean state).
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "a"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("use a: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	r = mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "sync"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("sync clean: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(strings.ToLower(r.Stdout), "synced a") {
		t.Errorf("stdout missing 'synced a': %q", r.Stdout)
	}
}
